File: sleddingHills.js
function sleddingHills(enjoyability, maxRuns) {
    let total = 0;
    
    // sort array
    enjoyability = enjoyability.sort((a,b)=> a-b);
    
    while (maxRuns > 0) {
        let max = enjoyability[enjoyability.length-1];
        
        if (max <= 0) break;
        total += max;
        
        // remove last one
        enjoyability.pop();
        
        let position = binarySearchPosition(enjoyability, max - 1);
        
        // insert element (max-1) into sorted enjoyability
        if(position == -1){
            enjoyability.push(max - 1);
        }else enjoyability.splice(position, 0, max - 1);
        
        maxRuns--;
    }

    return total;
}

// find the first element greater than target in a sorted array
function binarySearchPosition(arr, target){
    let start = 0, end = arr.length-1;
    let ans = -1;
    
    while(start <= end){
        let mid =  Math.floor((start + end)/2);
        
        if(arr[mid] <= target){
            start = mid + 1;
        }else{
            ans = mid;
            end = mid - 1;
        }
    }
    
    return ans;
}

File: sleddingHills.java
int sleddingHills(int[] enjoyability, int maxRuns) {
    
    int total = 0;
    // sort array
    Arrays.sort(enjoyability);
    ArrayList<Integer> list = new ArrayList<Integer>();
    
    for(int i = 0;i< enjoyability.length;i++){
        list.add(enjoyability[i]);
    }
    
    while(maxRuns > 0){
        int max = list.get(list.size()-1);
        int next = max - 1;
        if(max <=0){
            break;
        }
        
        total += max;
        
        // remove last element in list
        list.remove(list.size()-1);
        
        // insert (max-1) into list to maintain a sorted list
        int position = binarySearchPosition(list, next);
        
        if(position == -1){
            list.add(next);
        }else{
            list.add(position, next);
        }
        
        maxRuns -=1;
    }
    
    
    return total;
}

int binarySearchPosition(ArrayList<Integer> list, int target){
    int start = 0, end = list.size()-1;
    int ans = -1;
    
    while(start <= end){
        int mid = ((start + end)/2);
        
        if(list.get(mid) <= target){
            start = mid + 1;
        }else{
            ans = mid;
            end = mid - 1;
        }
    }
    
    return ans;
}